contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,396 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.20;
+
+import "../libraries/SwapIntention.sol";
+import {IQuoter} from "../interfaces/IQuoter.sol";
+import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+
+contract Quoter is IQuoter {
+    using PoolIdLibrary for PoolKey;
+    using Hooks for IHooks;
+
+    /// @dev Transient storage variable used to check a safety condition in exact output swaps.
+    uint256 private amountOutCached;
+
+    // v4 Singleton contract
+    IPoolManager public immutable manager;
+
+    constructor(address _poolManager) {
+        manager = IPoolManager(_poolManager);
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInputSingle(ExactInputSingleParams memory params)
+        public
+        override
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+    {
+        try manager.lock(abi.encode(SwapInfo(SwapType.ExactInputSingle, abi.encode(params)))) {}
+        catch (bytes memory reason) {
+            return _handleRevertSingle(reason, params.poolKey);
+        }
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInput(ExactInputParams memory params)
+        external
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        try manager.lock(abi.encode(SwapInfo(SwapType.ExactInput, abi.encode(params)))) {}
+        catch (bytes memory reason) {
+            return _handleRevert(reason);
+        }
+    }
+
+    function quoteExactInputBatch(ExactInputSingleBatchParams memory params)
```

Comment:
we can do batch quotes via `multicall` instead for a cleaner interface, unless there are important benefits here that I'm missing?

Reply:
++ for multicall

Reply:
removed